### PR TITLE
HPCC-14603 Compress serialized stats format

### PR DESCRIPTION
**Pull request by ghalliday**
These helper functions may be useful in other situations...

Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith please review.

File: system/jlib/jbuff.cpp [hunk 1]
```diff
@@ -554,6 +554,18 @@ MemoryBuffer & MemoryBuffer::append(unsigned __int64 value)
...
+MemoryBuffer & MemoryBuffer::appendPacked(unsigned __int64 value)
+{
+    //Append bytes with the top bit set until the value is less than 0x80
+    while (value >= 0x80)
+    {
+        byte next = ((byte)value) | 0x80;
+        append(next);
+        value >>= 7;
+    }
+    return append((byte)value);
```
Comment (jakesmith) [inline system/jlib/jbuff.cpp]: Just checking, but worst case here is to append 10 bytes?
But approach makes sense where usually relatively small numbers are being serialized.

File: system/jlib/jbuff.cpp [hunk 2]
```diff
@@ -736,6 +748,34 @@ const byte * MemoryBuffer::readDirect(size32_t len)
...
+        if (!(next & 0x80))
+            break;
+        shift += 7;
+    }
+    return value;
+}
+
+MemoryBuffer &  MemoryBuffer::readPacked(unsigned & value)
+{
+    value = (unsigned)readPacked();
```
Comment (jakesmith) [inline system/jlib/jbuff.cpp]: Shouldn't this, in debug at least, impost a limit on number bytes read? i.e. if it reads >5 bytes it's overrun what could have possibly been serialized for a packed unsigned.

**Comment by jakesmith** [comment]
How much this gain on a large graph?
Since it was being LZW compressed anyway.

@ghalliday - a couple of minor question/points.

File: system/jlib/jbuff.cpp (hunk 1 above)
Comment (ghalliday) [inline system/jlib/jbuff.cpp]: True. I could avoid spilling to 10 bytes If I special cased the last byte

**Comment by ghalliday** [comment]
This is not used in the serialised stats. It is only the runtime classes which serialise information from the slaves to the master.

File: system/jlib/jbuff.cpp (hunk 2 above)
Comment (ghalliday) [inline system/jlib/jbuff.cpp]: I could add an assert on the value.

**Comment by ghalliday** [comment]
@jakesmith is that sufficient?

**Comment by jakesmith** [comment]
@ghalliday - yes that additional check adds some peace of mind.

Looks good to merge to me.

**Comment by ghalliday** [comment]
@richardkchapman

**Pull request merged**